extensions/wikia/SitemapXml/SitemapHooks.class.php

```diff
@@ -6,10 +6,72 @@ class SitemapHooks {
 	 * Disable title redirects for sitemaps, as we want to use the original request url rather than Special:Sitemap.
 	 */
 	public static function onBeforeTitleRedirect( WebRequest $request, Title $title ): bool {
-		if ( $title->isSpecial( 'Sitemap' ) ) {
+		if ( self::isSitemapTitle( $title ) ) {
 			return false;
 		}
 		return true;
 	}
 
+	public static function onTestCanonicalRedirect( WebRequest $request, Title $title, OutputPage $output ) {
+		global $wgServer, $wgScriptPath, $wgScript, $wgArticlePath;
+
+		if ( !self::isSitemapTitle( $title ) || WebRequest::detectProtocol() !== 'http' ) {
+			return true;
+		}
+
+		$currentHost = parse_url( WikiFactoryLoader::getCurrentRequestUri( $_SERVER, true, true ), PHP_URL_HOST );
+
+		// We only want to deal with plain HTTP versions of the wikia.com domain
+		if ( !self::shouldCancelRedirect( $currentHost, $title ) ) {
+			return true;
+		}
+
+		$output->cancelRedirect();
+
+		$request->response()->header( 'X-Sitemaps-Redirect-Cancelled: 1' );
+
+		// Override the wgServer so the sitemap uses host from the current request.
+		// This needs to be done because WFL used the address from city_url
+		$wgServer = "http://{$currentHost}";
+		// Reset script path and related variables on language path wikis
+		// so the URLs in the sitemap are correct
+		$wgScriptPath = '';
+		$wgScript = '/index.php';
+		$wgArticlePath = '/wiki/$1';
+		return false;
+	}
+
+	private static function shouldCancelRedirect( $currentHost, Title $targetTitle ): bool {
```

Comment:
This method is hard to follow. What do you think about adding a unit test for it? Would be easier to reason about it while looking at the test cases.

Reply:
Good point, I've added a unit test for the hook method, which will cover the private method logic.